helion/_compiler/matmul_utils.py

```diff
@@ -0,0 +1,299 @@
+    dtypes for fused accumulation when providing `acc`.
+    """
+
+    parts = ["tl.dot({lhs}, {rhs}"]
+    if acc is not None:
+        parts.append(", acc={acc}")
+    if input_precision is not None:
+        parts.append(f", input_precision='{input_precision}'")
+    if out_dtype is not None:
+        parts.append(f", out_dtype={triton_type(out_dtype)}")
+    parts.append(")")
+    tpl = "".join(parts)
+    kwargs: dict[str, ast.AST] = {"lhs": lhs, "rhs": rhs}
+    if acc is not None:
+        kwargs["acc"] = acc
+    return expr_from_string(tpl, **kwargs)
+
+
+def emit_tl_dot_with_padding(
+    lhs: ast.AST,
+    rhs: ast.AST,
+    acc: ast.AST | None,
+    lhs_dtype: torch.dtype,
+    rhs_dtype: torch.dtype,
+    *,
+    m: object,
+    n: object,
+    k: object,
+    shape_str_fn: Callable[[list[object]], str] | None = None,
+    acc_dtype: torch.dtype | None = None,
+    input_precision: str | None = None,
+    out_dtype: torch.dtype | None = None,
+) -> ast.AST:
+    """Emit a tl.dot operation with automatic padding for small dimensions.
+
+    This function handles matrix multiplication when dimensions are below hardware
+    requirements by automatically padding the matrices to meet minimum size constraints
+    and then unpadding the result.
+
+    Args:
+        lhs: Left-hand side matrix AST expression
+        rhs: Right-hand side matrix AST expression
+        acc: Optional accumulator AST expression for fused multiply-accumulate
+        lhs_dtype: PyTorch dtype of the left operand
+        rhs_dtype: PyTorch dtype of the right operand
+        m: M dimension (rows of lhs) - can be symbolic or concrete value
+        n: N dimension (cols of rhs) - can be symbolic or concrete value
+        k: K dimension (cols of lhs/rows of rhs) - can be symbolic or concrete value
+        shape_str_fn: Optional function to generate shape strings from dimension lists
+        acc_dtype: Optional dtype to cast the result to (when not fusing accumulator)
+        input_precision: Optional precision for the dot operation (e.g., 'tf32')
+        out_dtype: Optional output dtype for the dot operation
+
+    Returns:
+        AST expression for the (potentially padded and unpadded) dot product result
+
+    Raises:
+        DimensionRequirementError: If K dimension is below minimum requirement (cannot be padded)
+
+    Notes:
+        - M and N dimensions can be padded if below hardware requirements
+        - K dimension cannot be padded and will raise an error if too small
+        - When padding is applied, the result is automatically unpadded using sum reductions
+    """
+    device = CompileEnvironment.current().device
+    min_m, min_n, min_k = min_dot_size(device, lhs_dtype, rhs_dtype)
+
+    # Try to get concrete values for comparisons
+    m_val = int(m) if isinstance(m, int) else None
+    n_val = int(n) if isinstance(n, int) else None
+    k_val = int(k) if isinstance(k, int) else None
```

Comment:
Merge this with the conversion logic in the caller an pass SymInts to this function.